### PR TITLE
Zak acr

### PR DESCRIPTION
**Pull request by cshoebridge**
merge conflicts resolved

### CODE REVIEW

**Comment by cshoebridge** [comment]
Should now be ready with all the good changes

File: LION/models/learned_regularizer/ACR.py [hunk 1]
```diff
@@ -61,7 +63,7 @@ class L2net(nn.Module):
     def __init__(self):
         super(L2net, self).__init__()
 
-        self.l2_penalty = nn.Parameter((-9.0) * torch.ones(1))
+        self.l2_penalty = nn.Parameter((-36.0) * torch.ones(1))
```
Comment (AnderBiguri) [inline LION/models/learned_regularizer/ACR.py:66]: lets make this value an input for now. We'll need @Zakobian to weigh in if there is a reasonable way to estimate this value.

File: LION/models/learned_regularizer/ACR.py [hunk 2]
```diff
@@ -71,58 +73,32 @@ def forward(self, x):
...
-        if self.model_parameters.L2net:
-            out = out + self.L2net(x)
-        return out
-
-
-class ACR(LIONmodel.LIONmodel):
-    def __init__(self, model_parameters: LIONParameter = None):
-
-        super().__init__(model_parameters)
-
```
Comment (AnderBiguri) [inline LION/models/learned_regularizer/ACR.py:125]: I wrote this from Subhos code. I can't see it anywere (maybe Im blind) here. @Zakobian did you use this SBF in your code? why does your version of ACR not have it? or is it somewhere else that I am missing?

File: LION/models/learned_regularizer/ACR.py [hunk 3]
```diff
@@ -175,44 +173,47 @@ def initialize_weights(self, min_val=0.0, max_val=0.001):
...
-            }"""
+            @inproceedings{mukherjee2024data,
+            title={Data-Driven Convex Regularizers for Inverse Problems},
+            author={Mukherjee, S and Dittmer, S and Shumaylov, Z and Lunz, S and {\"O}ktem, O and Sch{\"o}nlieb, C-B},
+            booktitle={ICASSP 2024-2024 IEEE International Conference on Acoustics, Speech and Signal Processing (ICASSP)},
+            pages={13386--13390},
+            year={2024},
+            organization={IEEE}
+            }
+            """
```
Comment (AnderBiguri) [inline LION/models/learned_regularizer/ACR.py:216]: I would argue the appropiate citation is this:

https://scholar.google.co.uk/scholar?cluster=811116575160518931&hl=en&as_sdt=2005

File: LION/models/learned_regularizer/AR.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+class ARNetworkParams(ModelParams):
+    def __init__(
+        self,
+        n_channels: int,
+    ):
+        super().__init__(model_input_type=ModelInputType.IMAGE)
+        self.n_channels = n_channels
+
+
+class ARNetwork(LIONmodel):
```
Comment (AnderBiguri) [inline LION/models/learned_regularizer/AR.py:21]: @Zakobian What do you think about naming? @cshoebridge and I were discussing how technically the model is not "AR", the framework for training is an AR, so we called this an ARNetwork. But I feel that maybe that is misleading and people call the netowkr also AR? Any opinions on naming?

File: LION/models/learned_regularizer/AR.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            title={Data-Driven Convex Regularizers for Inverse Problems},
+            author={Mukherjee, S and Dittmer, S and Shumaylov, Z and Lunz, S and {\"O}ktem, O and Sch{\"o}nlieb, C-B},
+            booktitle={ICASSP 2024-2024 IEEE International Conference on Acoustics, Speech and Signal Processing (ICASSP)},
+            pages={13386--13390},
+            year={2024},
+            organization={IEEE}
+            }
+            """
+            print(string)
+        else:
```
Comment (AnderBiguri) [inline LION/models/learned_regularizer/AR.py:80]: https://scholar.google.co.uk/scholar?cluster=3594915696133260277&hl=en&as_sdt=2005&sciodt=0,5

File: LION/models/post_processing/FBPConvNet.py [hunk 6]
```diff
@@ -188,7 +193,7 @@ def __init__(
 
     @staticmethod
     def default_parameters():
-        FBPConvNet_params = LIONParameter()
+        FBPConvNet_params = FBPConvNetParams()
         FBPConvNet_params.down_1_channels = [1, 64, 64, 64]
         FBPConvNet_params.down_2_channels = [64, 128, 128]
         FBPConvNet_params.down_3_channels = [128, 256, 256]
```
Comment (AnderBiguri) [inline LION/models/post_processing/FBPConvNet.py:199]: move these to FBConvNetParams

File: LION/optimizers/ARsolver.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+        if dataloader is not None:
+            residual = 0.0
+            for _, (sino, target) in enumerate(dataloader):
+                residual += torch.linalg.vector_norm(
+                    self.AT(self.A(target) - sino), dim=(2, 3)
+                ).mean()
+                # residual += torch.sqrt(((self.AT(self.A(target) - data))**2).sum())
+            self.alpha = residual / len(dataloader.dataset)
+        if self.verbose:
+            print("Estimated alpha: " + str(self.alpha))
```
Comment (AnderBiguri) [inline LION/optimizers/ARsolver.py:69]: @Zakobian is this the power method for the operator norm? or something else?

File: LION/optimizers/ARsolver.py [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+                f"Testing loss mean: {test_loss.mean()} - Testing loss std: {test_loss.std()}"
+            )
+
+        # return to train if it was in train
+        if status:
+            self.model.train()
+
+        return test_loss
+
+    def wgan_loss(self, sino_batch, target_batch):
```
Comment (AnderBiguri) [inline LION/optimizers/ARsolver.py:216]: @Zakobian is "WGAN-loss" a name _too generic_? It is _a_ WGAN loss, of course, but would someone mistake this for something else? or I am overworrying about silly things?

File: LION/optimizers/ARsolver.py [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+            @inproceedings{mukherjee2024data,
+            title={Data-Driven Convex Regularizers for Inverse Problems},
+            author={Mukherjee, S and Dittmer, S and Shumaylov, Z and Lunz, S and {\"O}ktem, O and Sch{\"o}nlieb, C-B},
+            booktitle={ICASSP 2024-2024 IEEE International Conference on Acoustics, Speech and Signal Processing (ICASSP)},
+            pages={13386--13390},
+            year={2024},
+            organization={IEEE}
+            }
+            """
+            print(string)
```
Comment (AnderBiguri) [inline LION/optimizers/ARsolver.py:279]: I'd add here the Sebastian Lunz paper (AR).

File: LION/optimizers/LIONsolver.py [hunk 10]
```diff
@@ -141,46 +155,77 @@ def set_validation(
...
+    def set_loading(self, load_folder: str | pathlib.Path, do_load: bool = False):
+        if isinstance(load_folder, str):
+            load_folder = pathlib.Path(load_folder)
+        if not load_folder.is_dir():
+            raise ValueError(
+                f"Save folder '{load_folder}' is not a directory, failed to set saving."
+            )
+
+        self.load_folder = load_folder
+        self.do_load_checkpoint = do_load
```
Comment (AnderBiguri) [inline LION/optimizers/LIONsolver.py:192]: We were going to change this. Is this here because you haven't merged main here yet?

File: LION/models/learned_regularizer/ACR.py (hunk 1 above)
Comment (cshoebridge) [reply LION/models/learned_regularizer/ACR.py:66]: Another question with this, do we want these small modules to be LionModels? I'd suggest no, because as I view them, they're just extracting a bit of code form inside the ACR network. Doesn't seem grand enough

File: LION/optimizers/LIONsolver.py (hunk 10 above)
Comment (cshoebridge) [reply LION/optimizers/LIONsolver.py:192]: So do you want to just entirely do away with set_loading and have set_checkpointing have a signature like (checkpoint fname, checkpoint_freq, do_load, save_folder, load_folder)?

File: LION/models/learned_regularizer/ACR.py (hunk 2 above)
Comment (Zakobian) [reply LION/models/learned_regularizer/ACR.py:125]: SFB was in the ACR just for consistency check - not really part of the method

File: LION/models/learned_regularizer/AR.py (hunk 4 above)
Comment (Zakobian) [reply LION/models/learned_regularizer/AR.py:21]: I would say that AR is indeed a training framework, but I would say that ambiguity in calling both the training framework and the method is ok here, since the improvements on the model so far have primarily been network architecture based, thus distinguishing themselves from AR? Otherwise they are all technically AR. Does this make sense?

File: LION/optimizers/ARsolver.py (hunk 7 above)
Comment (Zakobian) [reply LION/optimizers/ARsolver.py:69]: Nope, this is estimating the regularisation parameter

File: LION/optimizers/ARsolver.py (hunk 8 above)
Comment (Zakobian) [reply LION/optimizers/ARsolver.py:216]: Id say this is overworrying, cause it is the WGAN loss

File: LION/models/learned_regularizer/ACR.py (hunk 1 above)
Comment (AnderBiguri) [reply LION/models/learned_regularizer/ACR.py]: @cshoebridge currently there is a serious issue with saving/loading that does not allow a LIONmodel inside a LIONmodel to be correctly saved. So for now, definetly no.

File: LION/models/learned_regularizer/ACR.py (hunk 2 above)
Comment (AnderBiguri) [reply LION/models/learned_regularizer/ACR.py]: @Zakobian so when you train it etc, you don't use it? 

What do you mean "consistency check"?

File: LION/models/learned_regularizer/AR.py (hunk 4 above)
Comment (AnderBiguri) [reply LION/models/learned_regularizer/AR.py:21]: So then maybe lets call it AR @cshoebridge

File: LION/optimizers/LIONsolver.py (hunk 10 above)
Comment (AnderBiguri) [reply LION/optimizers/LIONsolver.py]: yup! do it that way, then I will fiddle with it later on if needed

File: LION/optimizers/ARsolver.py (hunk 8 above)
Comment (cshoebridge) [reply LION/optimizers/ARsolver.py:216]: leave as is then @AnderBiguri
